### PR TITLE
Description: Automate monthly change for PL sandbox account passwords Issue: https://khanacademy.atlassian.net/browse/DIST-4092

### PR DESCRIPTION
**Pull request by rgagarinr**
Summary:
This is a cron job scheduled to run monthly Cypress/Python scripts for updating demo passwords in the PL Demo District.

Automate monthly change for PL sandbox account passwords.
Issue: DIST-4092

## Test plan:
Successfully ran the script manually on the Jenkins server.

### CODE REVIEW

File: jobs/demo-district-update-pass.groovy [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+).addStringParam(
+   "SLACK_CHANNEL",
+   "The slack channel to which to send failure alerts.",
+   "#demo-district-logs"
+
+).addStringParam(
+   "REVISION_DESCRIPTION",
+   """Set by the buildmaster to give a more human-readable description
+of the GIT_REVISION, especially if it is a commit rather than a branch.
+Defaults to GIT_REVISION.""",
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:31]: Looks like this comment isn't true.  Should it have a default value (on line 32) at all?

File: jobs/demo-district-update-pass.groovy [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+).addStringParam(
+   "TEST_RETRIES",
+   """How many retry attempts to use. By default is 1.""",
+   "1"
+
+).apply()
+
+// We use the build name as a unique identifier for user notifications. 
+BUILD_NAME = "build demo-district-password-update #${env.BUILD_NUMBER} (${params.REVISION_DESCRIPTION})"
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:42]: This is overkill for this job -- only one of these ever runs at a time.  You can look at other similar periodic jobs, like update-devserver-static-images.groovy, to see what kind of setup they do.

(I'd get rid of REVISION_DESCRIPTION entirely as well.)

**Review by csilvers — CHANGES_REQUESTED** [review]
The overall logic looks good to me!  But some work is going to be required to set this up to run successfully in a way that doesn't stomp on other jobs.

File: jobs/demo-district-update-pass.groovy [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+   """How many retry attempts to use. By default is 1.""",
+   "1"
+
+).addCronSchedule(
+   'H H 1 * *'        // Run on the first day of every month
+
+).apply()
+
+// We use the build name as a unique identifier for user notifications. 
+BUILD_NAME = "build demo-district-password-update #${env.BUILD_NUMBER}";
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:38]: I still don't think you need this -- the default BUILD_NAME is fine for you.

File: jobs/demo-district-update-pass.groovy [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                         \"screenshotOnRunFailure\":false, \"video\":false}"
+   ];
+
+   dir('webapp/services/static') {        
+      exec(runCypressArgs); 
+      sh '/usr/bin/python3 /home/ubuntu/webapp-workspace/webapp/services/static/javascript/districts-package/__e2e-tests__/update-psw.py'
+   }
+}
+
+onWorker("ka-test-ec2", '4d') {
```
[the remark this answers is not in the document]
Comment (csilvers) [reply jobs/demo-district-update-pass.groovy:71]: Still wondering about this.

File: jobs/demo-district-update-pass.groovy [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+def _setupWebapp() {
+   kaGit.safeSyncToOrigin("git@github.com:Khan/webapp", params.CYPRESS_GIT_REVISION);
+   dir("webapp/services/static") {
+      sh("yarn install");
+   }
+}
+
+def runScript() {
+   def tempDir = "webapp/services/static/javascript/districts-package/__e2e-tests__";
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:48]: What makes this a "temp" dir?  This is a directory in our source tree, what's temporary about it?

File: jobs/demo-district-update-pass.groovy [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+   def runCypressArgs = ["yarn",
+                         "cypress",
+                         "run", 
+                         "--spec",
+                         "javascript/districts-package/__e2e-tests__/change-pswd.spec.ts",
+                         "--browser=electron",
+                         "-c {\"retries\":${params.TEST_RETRIES}, \
+                         \"screenshotOnRunFailure\":false, \"video\":false}"
+   ];
+   def dockerBuild = ["docker", "build", "-t", "update-pass", "."];
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:58]: I don't see a Dockerfile in this directory.  Is that being added in another PR?

**Review by csilvers — COMMENTED** [review]
Getting much closer!  But I'm not comfortable approving it until I understand more what's going on with the Dockerfile.

**Comment by csilvers** [comment]
> @[rgagarinr](https://github.com/rgagarinr) rgagarinr requested a review from [csilvers](https://github.com/csilvers)

Just to be clear, the things I'm waiting on now:
1. Something pointing to this dockerfile that the job is using -- like the PR it's written in, or something
2. Some discussion of why you're running on a test-worker and not on master.

**Comment by rgagarinr** [comment]
> > @[rgagarinr](https://github.com/rgagarinr) rgagarinr requested a review from [csilvers](https://github.com/csilvers)
> 
> Just to be clear, the things I'm waiting on now:
> 
> 1. Something pointing to this dockerfile that the job is using -- like the PR it's written in, or something
> 2. Some discussion of why you're running on a test-worker and not on master.

Sorry, I didn't submit my previous comments 🙇 they were in a pending state. Thanks again for reviewing @csilvers !
1) Here is a [dockerfile](https://github.com/Khan/webapp/commit/a1f8b7c12714b6887bcd32b4366e89fcf1c492a1#diff-ef08d52616c7afe9a5f5f4acc9a2e0f23813dc57cab71dde6e558198610ce634) from the db brach. 
2) Instead of on `master`? I tried it, but there is no `master`. What about `big-test-worker`? I'm okay with any worker that is safer and faster.

File: jobs/demo-district-update-pass.groovy (hunk 6 above)
Comment (csilvers) [reply jobs/demo-district-update-pass.groovy:58]: Coolio.  I think a dockerfile is probably overkill here, but it should definitely work!

File: jobs/demo-district-update-pass.groovy [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      exec(runCypressArgs); 
+   }
+
+   dir(workDir) {
+      exec(dockerBuild); 
+      exec(dockerRun); 
+   }
+}
+
+onWorker("ka-test-ec2", '4d') {
```
Comment (csilvers) [inline jobs/demo-district-update-pass.groovy:71]:
```suggestion
onMaster('4d') {
```

**Review by csilvers — CHANGES_REQUESTED** [review]
Change to onMaster and I think we're good to go!

File: jobs/demo-district-update-pass.groovy (hunk 7 above)
Comment (rgagarinr) [reply jobs/demo-district-update-pass.groovy:71]: Thanks, I'll change it to OnMaster!

Comment (rgagarinr) [reply jobs/demo-district-update-pass.groovy:71]: @csilvers Seems like `OnMaster` is not [suitable](https://jenkins.khanacademy.org/job/misc/job/demo%20district/job/update-psw/41/console) for our (e2e) automation tasks, that's why we have been using the `ka-test-ec2` worker.

**Review by csilvers — APPROVED** [review]
Ah, good point.  OK, I'd add a comment to that effect ("we can't run this on master because jenkin-server doesn't have cypress deps installed"), and then you're good to go!

**Review by csilvers — APPROVED** [review]

**Pull request merged**